src/Parallel/GlobalCache.hpp

```diff
@@ -634,16 +440,48 @@ template <typename GlobalCacheTag, typename Function, typename... Args>
 void GlobalCache<Metavariables>::mutate(const std::tuple<Args...>& args) {
   (void)Parallel::charmxx::RegisterGlobalCacheMutate<
       Metavariables, GlobalCacheTag, Function, Args...>::registrar;
-  if (mutable_global_cache_proxy_is_set()) {
-    // charm-aware version: Mutate the variable on all PEs on this node.
-    for (auto pe = CkNodeFirst(CkMyNode());
-         pe < CkNodeFirst(CkMyNode()) + CkNodeSize(CkMyNode()); ++pe) {
-      mutable_global_cache_proxy_[pe].template mutate<GlobalCacheTag, Function>(
-          args);
-    }
-  } else {
-    // version that bypasses proxies.  Just call the function.
-    mutable_global_cache_->template mutate<GlobalCacheTag, Function>(args);
+  using tag = MutableCacheTag<GlobalCache_detail::get_matching_mutable_tag<
+      GlobalCacheTag, Metavariables>>;
+
+  // Do the mutate.
+  std::apply(
+      [this](const auto&... local_args) {
+        // First mutex is for value of mutable tag
+        std::mutex& mutex = tuples::get<MutexTag<tag>>(mutexes_).first;
+        // Scoped for lock guard
+        {
+          const std::lock_guard<std::mutex> lock(mutex);
+          Function::apply(make_not_null(&(std::get<0>(
+                              tuples::get<tag>(mutable_global_cache_)))),
+                          local_args...);
+        }
+      },
+      args);
+
+  // A callback might call mutable_cache_item_is_ready, which might
+  // add yet another callback to the vector of callbacks.  We don't
+  // want to immediately invoke this new callback as it might just add another
+  // callback again (and again in an infinite loop). And we don't want to remove
+  // it from the vector of callbacks before it is invoked otherwise we could get
+  // a deadlock. Therefore, after locking it, we std::move the vector of
+  // callbacks into a temporary vector, clear the original vector, and invoke
+  // the callbacks in the temporary vector.
+  std::vector<std::unique_ptr<Callback>> callbacks{};
+  // Second mutex is for vector of callbacks
+  std::mutex& mutex = tuples::get<MutexTag<tag>>(mutexes_).second;
+  {
+    // Scoped for lock guard
+    const std::lock_guard<std::mutex> lock(mutex);
+    callbacks = std::move(std::get<1>(tuples::get<tag>(mutable_global_cache_)));
+    std::get<1>(tuples::get<tag>(mutable_global_cache_)).clear();
+    std::get<1>(tuples::get<tag>(mutable_global_cache_)).shrink_to_fit();
+  }
+
+  // Invoke the callbacks.  Any new callbacks that are added to the
+  // list (if a callback calls mutable_cache_item_is_ready) will be
+  // saved and will not be invoked here.
+  for (auto& callback : callbacks) {
+    callback->invoke();
```

Comment:
I think you need to hold the write lock for this entire function.  Here's the potential problem I see:
* We have two writers, A and B preparing to make independent changes to a cache object.
* We have an element E with a callback waiting for the change B is going to make.

Suppose the following sequence of events:
1. A mutates the object, copies the callback list, and starts the callback for element E.
2. E checks the current value and determines it is not ready.
3. B mutates the object, copies the empty callback list, and returns with nothing to do.
4. E returns a new callback, which is added to the callback list.
5. A returns.

We now have E waiting on a change that has already happened.